src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/SafeHandles.Windows.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
+
+namespace System.DirectoryServices.Protocols
+{
+    internal sealed class BerSafeHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+        internal BerSafeHandle() : base(true)
+        {
+            SetHandle(Wldap32.ber_alloc(1));
+            if (handle == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+        }
+
+        internal BerSafeHandle(berval value) : base(true)
+        {
+            SetHandle(Wldap32.ber_init(value));
+            if (handle == IntPtr.Zero)
+            {
+                throw new BerConversionException();
+            }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Wldap32.ber_free(handle, 1);
+            return true;
+        }
+    }
+
+    internal sealed class ConnectionHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+        internal bool _needDispose = false;
+
+        internal ConnectionHandle() : base(true)
+        {
+            SetHandle(Wldap32.ldap_init(null, 389));
+
+            if (handle == IntPtr.Zero)
+            {
```

Comment:
I feel like the marshaller bypasses ctors when it creates SafeHandles as a return value, (maybe I'm wrong and it just finds a ctor that looks like one it knows how to call? or maybe a default is required and I'm having a very bad recollection day).  I'm not sure this code actually prevents invalid handles.

A more canonical representation of this would be to declare the ldap_init call to return a SafeConnectionHandle (instead of IntPtr) and the caller should do

```C#
SafeConnectionHandle handle = LdapPal.Initialize(null, 389);
```

// In LdapPal.Windows.cs
```C#
internal static SafeConnectionHandle Initialize(string host, int port)
{
    SafeConnectionHandle handle = Interop.wldap32.ldap_init(host, port);

    if (handle.IsInvalid)
    {
        handle.Dispose();
        // this error handling code
    }

    return handle;
}
```

Reply:
> I feel like the marshaller bypasses ctors when it creates SafeHandles as a return value

Nope :)

e.g. try this:
```C#
using System;
using System.Runtime.InteropServices;

class Program
{
    static void Main() => GetStdHandle(-10);

    [DllImport("kernel32.dll", SetLastError = true)]
    private static extern MySafeHandle GetStdHandle(int nStdHandle);
}

class MySafeHandle : SafeHandle
{
    public MySafeHandle(string hello, string world) : base(IntPtr.Zero, ownsHandle: false) { }
    public override bool IsInvalid => true;
    protected override bool ReleaseHandle() => true;
}
```
You'll get an exception like this:
```
Unhandled Exception: System.MissingMethodException: .ctor
   at Program.GetStdHandle(Int32 nStdHandle)
   at Program.Main()
```